File: apps/code-runner/languages.py
"""Unified Language Configuration for the Sandbox Execution Engine."""
from typing import Dict, Optional
from pydantic import BaseModel


class LanguageConfig(BaseModel):
    id: str
    name: str
    extension: str
    default_filename: str
    docker_image: str
    compile_cmd: Optional[str] = None
    run_cmd: str
    timeout_seconds: float = 5.0
    memory_limit: str = "256m"
    pids_limit: int = 64


SUPPORTED_LANGUAGES: Dict[str, LanguageConfig] = {
    "python": LanguageConfig(
        id="python",
        name="Python",
        extension=".py",
        default_filename="main.py",
        docker_image="python:3.11-alpine",
        compile_cmd=None,
        run_cmd="python main.py",
        timeout_seconds=5.0,
        memory_limit="256m",
        pids_limit=64,
    ),
    "javascript": LanguageConfig(
        id="javascript",
        name="JavaScript (Node.js)",
        extension=".js",
        default_filename="index.js",
        docker_image="node:20-alpine",
        compile_cmd=None,
        run_cmd="node index.js",
        timeout_seconds=5.0,
        memory_limit="256m",
        pids_limit=64,
    ),
    "typescript": LanguageConfig(
        id="typescript",
        name="TypeScript",
        extension=".ts",
        default_filename="index.ts",
        docker_image="node:20-alpine",
        compile_cmd="npx -y esbuild index.ts --bundle --platform=node --outfile=dist.js",
        run_cmd="node dist.js",
        timeout_seconds=7.0,
        memory_limit="256m",
        pids_limit=64,
    ),
    "cpp": LanguageConfig(
        id="cpp",
        name="C++",
        extension=".cpp",
        default_filename="main.cpp",
        docker_image="gcc:13-alpine",
        compile_cmd="g++ -O2 -std=c++17 -o solution main.cpp",
        run_cmd="./solution",
        timeout_seconds=5.0,
        memory_limit="256m",
        pids_limit=64,
    ),
    "c": LanguageConfig(
        id="c",
        name="C",
        extension=".c",
        default_filename="main.c",
        docker_image="gcc:13-alpine",
        compile_cmd="gcc -O2 -std=c11 -o solution main.c",
        run_cmd="./solution",
        timeout_seconds=5.0,
        memory_limit="256m",
        pids_limit=64,
    ),
    "go": LanguageConfig(
        id="go",
        name="Go",
        extension=".go",
        default_filename="main.go",
        docker_image="golang:1.22-alpine",
        compile_cmd="go build -o solution main.go",
        run_cmd="./solution",
        timeout_seconds=5.0,
        memory_limit="256m",
        pids_limit=64,
    ),
    "rust": LanguageConfig(
        id="rust",
        name="Rust",
        extension=".rs",
        default_filename="main.rs",
        docker_image="rust:1.78-alpine",
        compile_cmd="rustc -O -o solution main.rs",
        run_cmd="./solution",
        timeout_seconds=7.0,
        memory_limit="256m",
        pids_limit=64,
    ),
    "java": LanguageConfig(
        id="java",
        name="Java",
        extension=".java",
        default_filename="Main.java",
        docker_image="eclipse-temurin:21-alpine",
        compile_cmd="javac Main.java",
        run_cmd="java Main",
        timeout_seconds=6.0,
        memory_limit="384m",
        pids_limit=64,
    ),
    "sql": LanguageConfig(
        id="sql",
        name="SQL (SQLite)",
        extension=".sql",
        default_filename="query.sql",
        docker_image="python:3.11-alpine",
        compile_cmd=None,
        run_cmd="sqlite3 db.sqlite < query.sql",
        timeout_seconds=4.0,
        memory_limit="256m",
        pids_limit=64,
    ),
}


def get_language_config(language_id: str) -> LanguageConfig:
    lang = language_id.lower().strip()
    if lang in SUPPORTED_LANGUAGES:
        return SUPPORTED_LANGUAGES[lang]
    raise ValueError(f"Unsupported language: '{language_id}'. Supported languages: {list(SUPPORTED_LANGUAGES.keys())}")

File: packages/types/src/languages.ts
export interface LanguageConfig {
  id: string;
  name: string;
  extension: string;
  defaultFileName: string;
  monacoLanguage: string;
  starterCode: string;
}

export const SUPPORTED_CODING_LANGUAGES: LanguageConfig[] = [
  {
    id: 'python',
    name: 'Python',
    extension: '.py',
    defaultFileName: 'main.py',
    monacoLanguage: 'python',
    starterCode: `def solve():\n    # Write your solution here\n    print("Hello, InterviewOS!")\n\nif __name__ == "__main__":\n    solve()\n`,
  },
  {
    id: 'javascript',
    name: 'JavaScript (Node.js)',
    extension: '.js',
    defaultFileName: 'index.js',
    monacoLanguage: 'javascript',
    starterCode: `function solve() {\n    // Write your solution here\n    console.log("Hello, InterviewOS!");\n}\n\nsolve();\n`,
  },
  {
    id: 'typescript',
    name: 'TypeScript',
    extension: '.ts',
    defaultFileName: 'index.ts',
    monacoLanguage: 'typescript',
    starterCode: `function solve(): void {\n    // Write your solution here\n    console.log("Hello, InterviewOS!");\n}\n\nsolve();\n`,
  },
  {
    id: 'java',
    name: 'Java',
    extension: '.java',
    defaultFileName: 'Main.java',
    monacoLanguage: 'java',
    starterCode: `public class Main {\n    public static void main(String[] args) {\n        // Write your solution here\n        System.out.println("Hello, InterviewOS!");\n    }\n}\n`,
  },
  {
    id: 'cpp',
    name: 'C++',
    extension: '.cpp',
    defaultFileName: 'main.cpp',
    monacoLanguage: 'cpp',
    starterCode: `#include <iostream>\n\nint main() {\n    // Write your solution here\n    std::cout << "Hello, InterviewOS!" << std::endl;\n    return 0;\n}\n`,
  },
  {
    id: 'c',
    name: 'C',
    extension: '.c',
    defaultFileName: 'main.c',
    monacoLanguage: 'c',
    starterCode: `#include <stdio.h>\n\nint main() {\n    // Write your solution here\n    printf("Hello, InterviewOS!\\n");\n    return 0;\n}\n`,
  },
  {
    id: 'go',
    name: 'Go',
    extension: '.go',
    defaultFileName: 'main.go',
    monacoLanguage: 'go',
    starterCode: `package main\n\nimport "fmt"\n\nfunc main() {\n    // Write your solution here\n    fmt.Println("Hello, InterviewOS!")\n}\n`,
  },
  {
    id: 'rust',
    name: 'Rust',
    extension: '.rs',
    defaultFileName: 'main.rs',
    monacoLanguage: 'rust',
    starterCode: `fn main() {\n    // Write your solution here\n    println!("Hello, InterviewOS!");\n}\n`,
  },
  {
    id: 'sql',
    name: 'SQL (SQLite)',
    extension: '.sql',
    defaultFileName: 'query.sql',
    monacoLanguage: 'sql',
    starterCode: `-- Write your SQL query here\nSELECT 'Hello, InterviewOS!' AS message;\n`,
  },
];

export const getLanguageConfig = (id: string): LanguageConfig => {
  return (
    SUPPORTED_CODING_LANGUAGES.find((l) => l.id.toLowerCase() === id.toLowerCase()) ||
    SUPPORTED_CODING_LANGUAGES[0]
  );
};
